File: cpp/Longest-Palindromic-Substring/Longest-Palindromic-Substring.cpp
/// Source : https://leetcode-cn.com/problems/longest-palindromic-substring/
/// Author : bryce
/// Time   : 2019-10-26

/*
由于字符串长度小于1000，因此我们可以用 O(n^2)O的算法枚举所有可能的情况。
首先枚举回文串的中心 ii，然后分两种情况向两边扩展边界，直到遇到不同字符为止:

回文串长度是奇数，则依次判断 s[i−k]==s[i+k],k=1,2,3,…
回文串长度是偶数，则依次判断 s[i−k]==s[i+k−1],k=1,2,3,…
如果遇到不同字符，则我们就找到了以 i 为中心的回文串边界。

时间复杂度分析：一共两重循环，所以时间复杂度是 O(n^2)。


*/

#include <iostream>
#include <string>
#include <cassert>

using namespace std;
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.size()<2) return s;
        int n = s.size(), maxlen = 0, start = 0;
        for (int i = 0; i < n - 1; i++ )
        {
            findPalindrome(s, i, i, start, maxlen);//奇数情况
            findPalindrome(s, i, i+1, start, maxlen);//偶数情况
        }
        return s.substr(start, maxlen);
    }
    
    void findPalindrome(string s, int left, int right, int& start, int& maxlen){//用&是通过引用访问实参 控制实参的
        while(left >=0 && s[left] == s[right]){
            left--;
            right++;
        }
        if (maxlen < right - left - 1){
            start = left + 1;
            maxlen = right - left - 1;
        }
    }
};

int main() {

    cout << Solution().longestPalindrome( "babad" )<<endl; 
    cout << Solution().longestPalindrome( "cbbd" )<<endl; 
    return 0;
}

File: python/Longest-Palindromic-Substring/Longest-Palindromic-Substring.py
#! /usr/bin/env python
#! -*- coding=utf-8 -*-
# Author: Bryce


class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        
        l=list(s)
        length = len(l)
        if length == 0:
            return ""
        if length == 1:
            return s
        temp=[]
        result = []
        i = 0
        ii = 0
        j = 1
        flag = True
        while ii < length:
            i = ii - 1
            j = ii + 1
            while j < length and l[ii] == l[j]:
                j = j + 1
                temp = l[ii:j]
            while (i>=0 and j<=length-1 and l[i] == l[j]):
                temp = l[i:j+1]
                i = i - 1
                j = j + 1
            if len(temp) > len(result):
                result = temp
            ii = ii + 1            
        if len(result) == 0:
            return l[0]
        return "".join(result)





class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        # 如果字符串长度小于2或者s等于它的倒序，则直接返回s
        if len(s) < 2 or s == s[::-1]:
            return s
        n = len(s)
        # 定义起始索引和最大回文串长度，odd奇，even偶
        start, maxlen = 0, 1
        # 因为i=0的话必然是不可能会有超过maxlen情况出现，所以直接从1开始
        for i in range(1, n):
            # 取i及i前面的maxlen+2个字符
            odd = s[i - maxlen - 1:i + 1]  # len(odd)=maxlen+2
            # 取i及i前面的maxlen+1个字符
            even = s[i - maxlen:i + 1]  # len(even)=maxlen+1
            if i - maxlen - 1 >= 0 and odd == odd[::-1]:
                start = i - maxlen - 1
                maxlen += 2
                continue
            if i - maxlen >= 0 and even == even[::-1]:
                start = i - maxlen
                maxlen += 1
        return s[start:start + maxlen]
